File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        ListNode* curr = nullptr; 
        ListNode* tracker1 = list1;
        ListNode* tracker2 = list2;
        ListNode* head; 
        if(list1 == nullptr){
            return list2; 
        }
        else if(list2 == nullptr){
            return list1; 
        }
        if(list1 -> val < list2 -> val){
            curr = list1; 
            head = curr; 
            tracker1 = list1 -> next;
        }
        else if(list1 -> val > list2 -> val || list1 -> val == list2 -> val){
            curr = list2; 
            head = curr; 
            tracker2 = list2 -> next;
        }
        while(tracker2 != nullptr && tracker1 != nullptr){
            if(tracker1 -> val < tracker2 -> val){
                curr -> next = tracker1; 
                tracker1 = tracker1 -> next;
            }
            else if(tracker2 -> val < tracker1 -> val || tracker2 -> val == tracker1 -> val){
                curr -> next = tracker2; 
                tracker2 = tracker2 -> next;
            }
            curr = curr -> next; 
        }
        if(tracker2 == nullptr){
            while(tracker1 != nullptr){
                curr -> next = tracker1; 
                tracker1 = tracker1 -> next; 
                curr = curr -> next; 
            }
        }
        if(tracker1 == nullptr){
            while(tracker2 != nullptr){
                curr -> next = tracker2; 
                tracker2 = tracker2 -> next; 
                curr = curr -> next; 
            }
        }
        return head; 
    }
};

File: Data Structures & Algorithms/reorder-linked-list/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        tracker = {}

        dummy = head
        answer = ListNode()
        tail = answer
        length = 0
        x = 0
        while dummy:
            tracker[x] = dummy
            length += 1
            x += 1
            dummy = dummy.next
        
        left = 0
        right = length - 1
        for i in range(length):
            if i%2 == 0:
                tail.next = tracker[left]
                left += 1
            else:
                tail.next = tracker[right]
                right -= 1
            tail = tail.next
        tail.next = None
        return None 
